cpp/dolfinx/io/ADIOS2Writers.h

```diff
@@ -867,6 +867,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
+          {(std::size_t)local_range_cell[0], (std::size_t)0},
+          {(std::size_t)num_cells_local, (std::size_t)num_dofs_per_cell});
+  engine.Put<std::int64_t>(local_topology, connectivity_out.data());
+
+  // Compute mesh permutations
+  topology->create_entity_permutations();
+  const std::vector<std::uint32_t>& cell_perm
+      = topology->get_cell_permutation_info();
+  adios2::Variable<std::uint32_t> local_cell_perm
+      = impl_adios2::define_variable<std::uint32_t>(
+          io, "CellPermutations", {(std::size_t)num_cells_global},
+          {(std::size_t)local_range_cell[0]}, {(std::size_t)num_cells_local});
+  engine.Put<std::uint32_t>(local_cell_perm, cell_perm.data());
+
+  engine.PerformPuts();
+}
+
+} // namespace impl_checkpoint
+
+template <std::floating_point T>
+class ADIOS2Checkpointer
+{
+public:
+  /// @privatesection
+  using Fd32 = fem::Function<float, T>;
+  using Fd64 = fem::Function<double, T>;
+  using Fc64 = fem::Function<std::complex<float>, T>;
+  using Fc128 = fem::Function<std::complex<double>, T>;
+  using U = std::vector<
+      std::variant<std::shared_ptr<const Fd32>, std::shared_ptr<const Fd64>,
+                   std::shared_ptr<const Fc64>, std::shared_ptr<const Fc128>>>;
+
+  /// @brief Create a ADIOS2 checkpointer for writing a mesh to file
+  ///
+  /// This format supports arbitrary degree meshes.
+  ///
+  /// @param[in] comm The MPI communicator to open the file on
+  /// @param[in] filename Name of output file
+  /// @param[in] mesh The mesh
+  /// @note This format support arbitrary degree meshes
+  ADIOS2Checkpointer(MPI_Comm comm, const std::filesystem::path& filename,
+                     std::shared_ptr<const mesh::Mesh<T>> mesh)
+      : _adios(std::make_unique<adios2::ADIOS>(comm)),
+        _io(std::make_unique<adios2::IO>(
+            _adios->DeclareIO("ADIOS2 checkpoint writer"))),
+        _engine(std::make_unique<adios2::Engine>(
+            _io->Open(filename, adios2::Mode::Write))),
+        _mesh(mesh), _u({}), _comm(comm)
+  {
+    // Define time independent attributes
+    auto topology = mesh->topology();
+    assert(topology);
+    if (topology->cell_types().size() > 1)
+      throw std::runtime_error("Multiple cell types not supported");
+    impl_adios2::define_attribute<int>(*this->_io, "CellType",
+                                       int(topology->cell_types()[0]));
+
+    const mesh::Geometry<T>& geometry = mesh->geometry();
+    if (geometry.cmaps().size() > 1)
+      throw std::runtime_error("Multiple coordinate maps not supported");
+    impl_adios2::define_attribute<int>(*this->_io, "Degree",
+                                       geometry.cmaps()[0].degree());
+    impl_adios2::define_attribute<int>(*this->_io, "LagrangeVariant",
+                                       int(geometry.cmaps()[0].variant()));
+  }
+
+  /// @brief Create a ADIOS2 checkpointer for reading a mesh from file
```

Comment:
Does this need to be a class? Would a function suffice?

Reply:
Maybe, it depends on how we want to do it when adding functions to the mix. have stand-alone functions for each of the operations, or having them interact in a class.